components/script/dom/element.rs

```diff
@@ -1691,6 +1691,20 @@ impl Element {
         }
     }
 
+    pub(crate) fn is_css_root_element(&self) -> bool {
+        // check if we're top of the tree that we're in
+        if *self.root_element() != *self {
+            return false;
+        }
+
+        // we need to be in a document tree but not in a shadow tree
+        if self.upcast::<Node>().is_in_a_shadow_tree() {
+            return false;
+        }
+
+        self.node.is_in_a_document_tree()
+    }
+
```

Comment:
Maybe it would make more sense to have a  `shadow_including_root_element` and `shadow_excluding_root_element`. "CSS root element" is not a specification concept AFAIK (correct me if I'm wrong), so we should avoid adding it.